src/hook.rs

```diff
@@ -0,0 +1,162 @@
+///
+/// # Example
+///
+/// Constructs a `HookTriggers` structure that tells Lua to call the hook after every instruction.
+/// Note the use of `Default`.
+///
+/// ```
+/// # use rlua::HookTriggers;
+/// # fn main() {
+/// let triggers = HookTriggers {
+///     after_counts: Some(1), ..Default::default()
+/// };
+/// # let _ = triggers;
+/// # }
+/// ```
+pub struct HookTriggers {
+    /// Before a function call.
+    pub calls: bool,
+    /// When Lua returns from a function.
+    pub returns: bool,
+    /// Before executing a new line, or returning from a function call.
+    pub lines: bool,
+    /// After a certain amount of instructions. When set to `Some(count)`, `count` is the number of
+    /// instructions to execute before calling the hook.
+    pub after_counts: Option<u32>,
+}
+
+impl HookTriggers {
+    /// Compute the mask to pass to `lua_sethook`.
+    pub(crate) fn mask(&self) -> c_int {
+        let mut mask: c_int = 0;
+        if self.calls { mask |= ffi::LUA_MASKCALL }
+        if self.returns { mask |= ffi::LUA_MASKRET }
+        if self.lines { mask |= ffi::LUA_MASKLINE }
+        if self.after_counts.is_some() { mask |= ffi::LUA_MASKCOUNT }
+        mask
+    }
+
+    /// Returns the `count` parameter to pass to `lua_sethook`, if applicable. Otherwise, zero is
+    /// returned.
+    pub(crate) fn count(&self) -> c_int {
+        self.after_counts.unwrap_or(0) as c_int
+    }
+}
+
+impl Default for HookTriggers {
+    fn default() -> Self {
+        HookTriggers {
+            calls: false,
+            returns: false,
+            lines: false,
+            after_counts: None
+        }
+    }
+}
+
+/// This callback is passed to `lua_sethook` and gets called whenever debug information is received.
+pub(crate) unsafe extern "C" fn hook_proc(state: *mut lua_State, ar: *mut lua_Debug) {
+    callback_error(state, || {
+        if ffi::lua_getinfo(state, cstr!("nSltu"), ar) == 0 {
+            rlua_panic!("lua_getinfo failed")
+        }
+
+        let extra = &mut *extra_data(state);
+
+        let debug = Debug {
+            name: ptr_to_str((*ar).name as *const i8),
+            namewhat: ptr_to_str((*ar).namewhat as *const i8),
+            what: ptr_to_str((*ar).what as *const i8),
+            source: ptr_to_str((*ar).source as *const i8),
+            curr_line: (*ar).currentline as u32,
+            line_defined: (*ar).linedefined as u32,
+            last_line_defined: (*ar).lastlinedefined as u32,
+            num_ups: (*ar).nups as u32,
+            num_params: (*ar).nparams as u32,
+            is_vararg: (*ar).isvararg == 1,
+            is_tailcall: (*ar).istailcall == 1,
+            short_src: str::from_utf8(mem::transmute((*ar).short_src.as_ref()))
+                .and_then(|r| Ok(Some(Cow::from(r))))
+                .unwrap_or(None),
+            _unused: ()
+        };
+
+        let cb = extra.hook_callback
+            .as_mut()
+            .expect("no hooks previously set; this is a bug");
```

Comment:
I think there should probably be an `rlua_expect` macro inside the library that prints consistent messages to match `rlua_assert` and `rlua_abort` and that `rlua_expect` should be used here and also everywhere that currently `.unwrap()` is used.  I also like that you say "this is a bug", and I think that all of the error macros in macros.rs should also say that.